system_setup/common/wsl_utils.py

```diff
@@ -92,3 +92,33 @@ def get_windows_locale():
     except OSError as e:
         log.info(windows_locale_failed_msg + e.strerror)
         return None
+
+
+def perform_wsl_shutdown(is_restart=False):
+    exec_wsl_map = {
```

Comment:
I don’t think we should hardcode the exe name here. This is not maintainable. It means that at each new releases, we need to add this element here, rerelease the oobe and so on. Can’t we get from the launcher side the current exe, with current arguments. and reexec this?

I think you tested it, but there is no chance that terminating the process prevents the second part (after the `;` to reexecute?)

Would `wsl.exec -d` bypass the launcher?

Reply:
> I don’t think we should hardcode the exe name here.

that would be true... I will modify this part

> I think you tested it, but there is no chance that terminating the process prevents the second part (after the ; to reexecute?)
>
> Would wsl.exec -d bypass the launcher?

What do you mean by these two questions? I am not sure I get what you mean... @didrocks

Reply:
> What do you mean by these two questions? I am not sure I get what you mean... @didrocks

* Is there a chance of a race between terminating the WSL instance (so this code) and the restart?
* wsl.exe -d "distro" is still using our launcher, correct?

Reply:
> Is there a chance of a race between terminating the WSL instance (so this code) and the restart?

No it wont, `;` separates the command and it will always execute first line (here is the `wsl --shutdown`) and then perform the restart

> wsl.exe -d "distro" is still using our launcher, correct?

No, I put this for fallback since not all ubuntu installations on wsl will have the executables; also I just forget the add the logic for the ones that do not have the 'WSL_DISTRO_NAME', those should be fallback directly to `wsl.exe`._